pyftpdlib/handlers.py

```diff
@@ -3457,23 +3459,50 @@ def __init__(self, conn, server, ioloop=None):
         def __repr__(self):
             return FTPHandler.__repr__(self)
 
+        def verify_certs_callback(self, connection, x509,
+                                  errnum, errdepth, ok):
+            if not ok:
+                self.log("Bad client certificate detected.")
+            else:
+                self.log("Client certificate is valid.")
+            return ok
+
+        def get_ssl_context(self):
+            if self.ssl_context is None:
+                self.ssl_context = self.validate_ssl_options()
+                if self.client_certfile is not None:
+                    from OpenSSL.SSL import VERIFY_CLIENT_ONCE
+                    from OpenSSL.SSL import VERIFY_FAIL_IF_NO_PEER_CERT
+                    from OpenSSL.SSL import VERIFY_PEER
+                    self.ssl_context.set_verify(VERIFY_PEER |
+                                                VERIFY_FAIL_IF_NO_PEER_CERT |
+                                                VERIFY_CLIENT_ONCE,
+                                                self.verify_certs_callback)
+            return self.ssl_context
+
         @classmethod
-        def get_ssl_context(cls):
-            if cls.ssl_context is None:
-                if cls.certfile is None:
-                    raise ValueError("at least certfile must be specified")
-                cls.ssl_context = SSL.Context(cls.ssl_protocol)
-                if cls.ssl_protocol != SSL.SSLv2_METHOD:
-                    cls.ssl_context.set_options(SSL.OP_NO_SSLv2)
-                else:
-                    warnings.warn("SSLv2 protocol is insecure", RuntimeWarning)
-                cls.ssl_context.use_certificate_chain_file(cls.certfile)
-                if not cls.keyfile:
-                    cls.keyfile = cls.certfile
-                cls.ssl_context.use_privatekey_file(cls.keyfile)
-                if cls.ssl_options:
-                    cls.ssl_context.set_options(cls.ssl_options)
-            return cls.ssl_context
+        def validate_ssl_options(cls):
```

Comment:
Why not put everything into `get_ssl_context` then or rename it to `_validate_ssl_context`? Later on you do:

```
ctx = TLS_FTPHandler.validate_ssl_options()
```

...which is a bit weird